src/agent/src/image_rpc.rs

```diff
@@ -33,11 +58,227 @@ pub struct ImageService {
     sandbox: Arc<Mutex<Sandbox>>,
 }
 
+#[derive(Serialize, Debug, Default, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct IndexDescriptor {
+    pub schema_version: u8,
+    pub manifests: Vec<OciDescriptor>,
+}
+
 impl ImageService {
     pub fn new(sandbox: Arc<Mutex<Sandbox>>) -> Self {
         Self { sandbox }
     }
 
+    fn build_oci_path(cid: &str) -> PathBuf {
+        let mut oci_path = PathBuf::from("/tmp");
+        oci_path.push(cid);
+        oci_path.push(IMAGE_OCI);
+        oci_path
+    }
+
+    fn handle_layer_data(
+        image_data: &ImageData,
+        image_manifest: &mut OciManifest,
+        crypto_config: &CryptoConfig,
+        oci_blob_path: &Path,
+    ) -> Result<()> {
+        for layer in image_data.layers.iter() {
+            let layer_digest = layer.clone().sha256_digest();
+
+            if layer.media_type == MEDIA_TYPE_LAYER_GZIP_ENC
+                || layer.media_type == MEDIA_TYPE_LAYER_ENC
+            {
+                if let Some(decrypt_config) = &crypto_config.decrypt_config {
+                    for layer_desc in image_manifest.layers.iter_mut() {
+                        if layer_desc.digest == layer_digest {
+                            let (layer_decryptor, _dec_digest) = decrypt_layer(
```

Comment:
Six levels of indentation is too much imho. Could you create a `handle_layer()` function or similar to reduce this a bit?

Please also consider adding some tests for this file as it has zero atm :sob:

- https://github.com/kata-containers/kata-containers/blob/main/docs/code-pr-advice.md#unit-tests

Reply:
I'll move this part in decrypt_layer_data function to reduce the level of indentation.